File: backend/app/dcf.py
"""
Discounted cash flow calculator — pure functions, no I/O.

Driver-agnostic 3-stage DCF: Stage 1 grows at `stage1_growth`, decaying by
`stage1_decay` per year, for `stage1_years`; an optional flat Stage 2 follows
for `stage2_years` (0 = skip — a plain 2-stage model); then a terminal value
via Gordon growth and/or an exit multiple (both are always computed so each
can be sanity-checked against the other, regardless of which one feeds the
headline implied price).

Works on any of three driver metrics:
  - fcf:        classic FCFF-style DCF; net-debt bridge optional (default on)
  - revenue:    for pre-FCF growth companies; terminal is exit-multiple only —
                perpetuity-growing raw revenue as if it were cash isn't
                economically coherent, so Gordon growth is not offered
  - net_income: for banks/insurers where FCF/capex isn't meaningful; already
                equity-level (post-interest), so the net-debt bridge is
                forced off regardless of the include_net_debt_bridge flag

Forward dilution: shares grow at `dilution_rate` for `dilution_years` then
flatten. Equity value is divided by the *diluted* (post-dilution) share
count — today's shareholders' claim on the terminal value is shrunk by the
same dilution the projection assumes, which is the standard simplified
treatment (vs. modeling year-by-year ownership).
"""

from dataclasses import dataclass, asdict
from typing import Literal, Optional

Driver = Literal["fcf", "revenue", "net_income"]
TerminalMethod = Literal["gordon", "exit_multiple"]


@dataclass
class ScenarioInputs:
    stage1_growth: float
    stage1_decay: float           # pp/year decline in growth during stage 1, e.g. 0.03 for 3pp/yr
    stage1_years: int
    stage2_growth: float = 0.0    # flat growth for stage 2; ignored if stage2_years == 0
    stage2_years: int = 0         # 0 = skip stage 2 (pure 2-stage: decay then terminal)
    terminal_growth: float = 0.025
    terminal_method: TerminalMethod = "gordon"
    exit_multiple: Optional[float] = None   # x of terminal-year driver value
    discount_rate: float = 0.09


@dataclass
class DCFInputs:
    driver: Driver
    starting_value: float           # TTM FCF / Revenue / Net Income, $
    shares_outstanding: float
    net_debt: float
    include_net_debt_bridge: bool   # forced off when driver == "net_income"
    dilution_rate: float            # annual, e.g. 0.02 for 2%/yr
    dilution_years: int
    bear: ScenarioInputs
    base: ScenarioInputs
    bull: ScenarioInputs


def _growth_path(s: ScenarioInputs) -> list[float]:
    path = []
    g = s.stage1_growth
    for _ in range(s.stage1_years):
        path.append(g)
        g -= s.stage1_decay
    if s.stage2_years > 0:
        path.extend([s.stage2_growth] * s.stage2_years)
    return path


def _implied_multiple_from_gordon(terminal_growth: float, discount_rate: float) -> Optional[float]:
    if discount_rate <= terminal_growth:
        return None
    return (1 + terminal_growth) / (discount_rate - terminal_growth)


def _implied_growth_from_multiple(exit_multiple: float, discount_rate: float) -> Optional[float]:
    denom = 1 + exit_multiple
    if denom == 0:
        return None
    return (exit_multiple * discount_rate - 1) / denom


def _run_scenario(driver: Driver, starting_value: float, shares: float, net_debt: float,
                   include_bridge: bool, dilution_rate: float, dilution_years: int,
                   s: ScenarioInputs) -> dict:
    growth_path = _growth_path(s)
    years = len(growth_path)
    if years == 0:
        raise ValueError("Stage 1 + Stage 2 years must total at least 1")
    if s.terminal_method == "gordon" and s.discount_rate <= s.terminal_growth:
        raise ValueError("Discount rate must exceed terminal growth rate for the Gordon growth method")
    if s.terminal_method == "exit_multiple" and not s.exit_multiple:
        raise ValueError("Exit multiple method requires exit_multiple to be set")

    value = starting_value
    rows = []
    pv_sum = 0.0
    for year, g in enumerate(growth_path, start=1):
        value = value * (1 + g)
        discount = (1 + s.discount_rate) ** year
        pv = value / discount
        pv_sum += pv
        rows.append({"year": year, "growth": round(g, 4), "value": round(value, 2), "pv": round(pv, 2)})

    terminal_value = rows[-1]["value"]  # last explicit-year value, before terminal growth/multiple applied

    terminal_value_gordon = None
    if s.discount_rate > s.terminal_growth:
        terminal_fcf = terminal_value * (1 + s.terminal_growth)
        terminal_value_gordon = terminal_fcf / (s.discount_rate - s.terminal_growth)

    terminal_value_exit = terminal_value * s.exit_multiple if s.exit_multiple else None

    chosen_terminal = terminal_value_gordon if s.terminal_method == "gordon" else terminal_value_exit
    pv_terminal = chosen_terminal / ((1 + s.discount_rate) ** years)

    enterprise_value = pv_sum + pv_terminal
    equity_value = enterprise_value - net_debt if include_bridge else enterprise_value

    diluted_shares = shares * ((1 + dilution_rate) ** dilution_years) if shares else None
    implied_price = equity_value / diluted_shares if diluted_shares else None

    return {
        "assumptions": asdict(s),
        "projection": rows,
        "terminal_value_gordon": round(terminal_value_gordon, 2) if terminal_value_gordon is not None else None,
        "terminal_value_exit_multiple": round(terminal_value_exit, 2) if terminal_value_exit is not None else None,
        "implied_exit_multiple_from_gordon": round(_implied_multiple_from_gordon(s.terminal_growth, s.discount_rate) or 0, 2) if terminal_value_gordon is not None else None,
        "implied_perpetuity_growth_from_exit_multiple": round(_implied_growth_from_multiple(s.exit_multiple, s.discount_rate), 4) if s.exit_multiple else None,
        "terminal_method_used": s.terminal_method,
        "pv_explicit": round(pv_sum, 2),
        "pv_terminal": round(pv_terminal, 2),
        "enterprise_value": round(enterprise_value, 2),
        "net_debt_bridge_applied": include_bridge,
        "equity_value": round(equity_value, 2),
        "diluted_shares": round(diluted_shares, 2) if diluted_shares else None,
        "implied_price": round(implied_price, 2) if implied_price is not None else None,
    }


def run_dcf(inputs: DCFInputs) -> dict:
    include_bridge = inputs.include_net_debt_bridge and inputs.driver != "net_income"

    def run(s: ScenarioInputs) -> dict:
        return _run_scenario(inputs.driver, inputs.starting_value, inputs.shares_outstanding,
                              inputs.net_debt, include_bridge, inputs.dilution_rate,
                              inputs.dilution_years, s)

    scenarios = {
        "driver": inputs.driver,
        "net_debt_bridge_applied": include_bridge,
        "bear": run(inputs.bear),
        "base": run(inputs.base),
        "bull": run(inputs.bull),
    }
    scenarios["sensitivity"] = _sensitivity_table(inputs, include_bridge)
    return scenarios


def _sensitivity_table(inputs: DCFInputs, include_bridge: bool) -> dict:
    """5x5 grid of implied price varying discount rate and terminal growth
    around the base case. Always uses the Gordon growth method for the grid,
    regardless of which terminal method the base case actually uses."""
    base = inputs.base
    discount_steps = [base.discount_rate + d for d in (-0.02, -0.01, 0, 0.01, 0.02)]
    tg_steps = [base.terminal_growth + d for d in (-0.01, -0.005, 0, 0.005, 0.01)]

    grid = []
    for r in discount_steps:
        row = []
        for tg in tg_steps:
            if r <= tg:
                row.append(None)
                continue
            s = ScenarioInputs(
                stage1_growth=base.stage1_growth, stage1_decay=base.stage1_decay, stage1_years=base.stage1_years,
                stage2_growth=base.stage2_growth, stage2_years=base.stage2_years,
                terminal_growth=tg, terminal_method="gordon", discount_rate=r,
            )
            result = _run_scenario(inputs.driver, inputs.starting_value, inputs.shares_outstanding,
                                    inputs.net_debt, include_bridge, inputs.dilution_rate,
                                    inputs.dilution_years, s)
            row.append(result["implied_price"])
        grid.append(row)

    return {
        "discount_rate_axis": [round(r, 4) for r in discount_steps],
        "terminal_growth_axis": [round(tg, 4) for tg in tg_steps],
        "grid": grid,
    }

File: frontend/lib/dcf.ts
// Client-side port of backend/app/dcf.py — same math, same field names, so
// live (in-browser) results and saved/historical (server-computed) results
// render through identical components. Runs synchronously on every input
// change: no network round trip, no debounce needed.

export type Driver = 'fcf' | 'revenue' | 'net_income'
export type TerminalMethod = 'gordon' | 'exit_multiple'

export interface ScenarioInputs {
  stage1_growth: number
  stage1_decay: number
  stage1_years: number
  stage2_growth: number
  stage2_years: number
  terminal_growth: number
  terminal_method: TerminalMethod
  exit_multiple: number | null
  discount_rate: number
}

export interface ProjectionRow {
  year: number
  growth: number
  value: number
  pv: number
}

export interface ScenarioResult {
  assumptions: ScenarioInputs
  projection: ProjectionRow[]
  terminal_value_gordon: number | null
  terminal_value_exit_multiple: number | null
  implied_exit_multiple_from_gordon: number | null
  implied_perpetuity_growth_from_exit_multiple: number | null
  terminal_method_used: TerminalMethod
  pv_explicit: number
  pv_terminal: number
  enterprise_value: number
  net_debt_bridge_applied: boolean
  equity_value: number
  diluted_shares: number | null
  implied_price: number | null
  error: string | null
}

export interface SensitivityResult {
  discount_rate_axis: number[]
  terminal_growth_axis: number[]
  grid: (number | null)[][]
}

export interface DcfResult {
  driver: Driver
  net_debt_bridge_applied: boolean
  bear: ScenarioResult
  base: ScenarioResult
  bull: ScenarioResult
  sensitivity: SensitivityResult
}

function growthPath(s: ScenarioInputs): number[] {
  const path: number[] = []
  let g = s.stage1_growth
  for (let i = 0; i < s.stage1_years; i++) {
    path.push(g)
    g -= s.stage1_decay
  }
  if (s.stage2_years > 0) {
    for (let i = 0; i < s.stage2_years; i++) path.push(s.stage2_growth)
  }
  return path
}

export function impliedMultipleFromGordon(terminalGrowth: number, discountRate: number): number | null {
  if (discountRate <= terminalGrowth) return null
  return (1 + terminalGrowth) / (discountRate - terminalGrowth)
}

export function impliedGrowthFromMultiple(exitMultiple: number, discountRate: number): number | null {
  const denom = 1 + exitMultiple
  if (denom === 0) return null
  return (exitMultiple * discountRate - 1) / denom
}

const ERROR_RESULT = (message: string): ScenarioResult => ({
  assumptions: {} as ScenarioInputs, projection: [], terminal_value_gordon: null,
  terminal_value_exit_multiple: null, implied_exit_multiple_from_gordon: null,
  implied_perpetuity_growth_from_exit_multiple: null, terminal_method_used: 'gordon',
  pv_explicit: 0, pv_terminal: 0, enterprise_value: 0, net_debt_bridge_applied: false,
  equity_value: 0, diluted_shares: null, implied_price: null, error: message,
})

export function runScenario(
  startingValue: number, shares: number, netDebt: number, includeBridge: boolean,
  dilutionRate: number, dilutionYears: number, s: ScenarioInputs,
): ScenarioResult {
  const path = growthPath(s)
  const years = path.length
  if (years === 0) return ERROR_RESULT('Stage 1 + Stage 2 years must total at least 1')
  if (s.terminal_method === 'gordon' && s.discount_rate <= s.terminal_growth) {
    return ERROR_RESULT('Discount rate must exceed terminal growth rate for the Gordon growth method')
  }
  if (s.terminal_method === 'exit_multiple' && !s.exit_multiple) {
    return ERROR_RESULT('Exit multiple method requires exit_multiple to be set')
  }

  let value = startingValue
  const rows: ProjectionRow[] = []
  let pvSum = 0
  for (let i = 0; i < path.length; i++) {
    const g = path[i]
    const year = i + 1
    value = value * (1 + g)
    const discount = Math.pow(1 + s.discount_rate, year)
    const pv = value / discount
    pvSum += pv
    rows.push({ year, growth: g, value, pv })
  }

  const terminalBase = rows[rows.length - 1].value
  const terminalValueGordon = s.discount_rate > s.terminal_growth
    ? (terminalBase * (1 + s.terminal_growth)) / (s.discount_rate - s.terminal_growth)
    : null
  const terminalValueExit = s.exit_multiple ? terminalBase * s.exit_multiple : null

  const chosenTerminal = s.terminal_method === 'gordon' ? terminalValueGordon : terminalValueExit
  if (chosenTerminal == null) return ERROR_RESULT('Could not compute a terminal value with the given inputs')

  const pvTerminal = chosenTerminal / Math.pow(1 + s.discount_rate, years)
  const enterpriseValue = pvSum + pvTerminal
  const equityValue = includeBridge ? enterpriseValue - netDebt : enterpriseValue
  const dilutedShares = shares ? shares * Math.pow(1 + dilutionRate, dilutionYears) : null
  const impliedPrice = dilutedShares ? equityValue / dilutedShares : null

  return {
    assumptions: s,
    projection: rows,
    terminal_value_gordon: terminalValueGordon,
    terminal_value_exit_multiple: terminalValueExit,
    implied_exit_multiple_from_gordon: terminalValueGordon != null ? impliedMultipleFromGordon(s.terminal_growth, s.discount_rate) : null,
    implied_perpetuity_growth_from_exit_multiple: s.exit_multiple ? impliedGrowthFromMultiple(s.exit_multiple, s.discount_rate) : null,
    terminal_method_used: s.terminal_method,
    pv_explicit: pvSum,
    pv_terminal: pvTerminal,
    enterprise_value: enterpriseValue,
    net_debt_bridge_applied: includeBridge,
    equity_value: equityValue,
    diluted_shares: dilutedShares,
    implied_price: impliedPrice,
    error: null,
  }
}

function sensitivityTable(
  startingValue: number, shares: number, netDebt: number, includeBridge: boolean,
  dilutionRate: number, dilutionYears: number, base: ScenarioInputs,
): SensitivityResult {
  const discountSteps = [-0.02, -0.01, 0, 0.01, 0.02].map((d) => base.discount_rate + d)
  const tgSteps = [-0.01, -0.005, 0, 0.005, 0.01].map((d) => base.terminal_growth + d)

  const grid: (number | null)[][] = discountSteps.map((r) =>
    tgSteps.map((tg) => {
      if (r <= tg) return null
      const s: ScenarioInputs = {
        stage1_growth: base.stage1_growth, stage1_decay: base.stage1_decay, stage1_years: base.stage1_years,
        stage2_growth: base.stage2_growth, stage2_years: base.stage2_years,
        terminal_growth: tg, terminal_method: 'gordon', exit_multiple: null, discount_rate: r,
      }
      const result = runScenario(startingValue, shares, netDebt, includeBridge, dilutionRate, dilutionYears, s)
      return result.error ? null : result.implied_price
    })
  )

  return { discount_rate_axis: discountSteps, terminal_growth_axis: tgSteps, grid }
}

export function runDcf(
  driver: Driver, startingValue: number, shares: number, netDebt: number, includeNetDebtBridge: boolean,
  dilutionRate: number, dilutionYears: number,
  bear: ScenarioInputs, base: ScenarioInputs, bull: ScenarioInputs,
): DcfResult {
  const includeBridge = includeNetDebtBridge && driver !== 'net_income'
  const run = (s: ScenarioInputs) => runScenario(startingValue, shares, netDebt, includeBridge, dilutionRate, dilutionYears, s)
  return {
    driver,
    net_debt_bridge_applied: includeBridge,
    bear: run(bear),
    base: run(base),
    bull: run(bull),
    sensitivity: sensitivityTable(startingValue, shares, netDebt, includeBridge, dilutionRate, dilutionYears, base),
  }
}
